File: src/dna2rna/dna2rna.py
import sys
from rna2fuun.rna2fuun import commands as rnacommands

class DNAList(list):
    def __init__(self, iterable):
        list.__init__(self, iterable)
        self.reverse()
        
    def __str__(self):
        return list.__str__(list(reversed(self)))

    def popfront(self, num=1):
##        print "popfront", num, len(self)
        if num == 1:
            return self.pop()
        elif num == 0:
            pass
        else:
            r = self[-num:]
            del self[-num:]
            r.reverse()
            return r
        
    def reflen(self, ref):
        return abs(ref.stop-ref.start)
    
    def reference(self, start, stop):
        return (start, stop)
    
    def getref(self, ref):
        ls = len(self)
        lr = list(reversed(self[ls - ref[1]: ls - ref[0]]))
        return lr
    def get(self, i):
        return self[-i-1]
    
    def prepend(self, iterable):
        self.extend(reversed(iterable))  
        
    def find(self, substr, i):
        if not substr:
            return -1
        ix = 0
        for ii in xrange(i, len(self)):
##            if ii % 1000 == 0:
##                print ii           
            if self[-ii-1] == substr[ix]:
##                print "match", ix,len(substr)
                ix += 1
                if ix == len(substr):
                    print "found!", ii
                    return ii
            else:
##                print "no match"
                ix = 0
        return -1

class NoMoreData(Exception):
    pass

# Extracts a pattern from the list 'dna', starting at position 'pos'.
# Returns a tuple containing the pattern and a new position after the
# consumed bases: (pattern, pos)
# The 'rna' parameter is also modified during the process.
def pattern(dna, rna):
##    print "Pattern:", dna
    p = []
    lvl = 0
    try:
        while dna:
##            print dna.get(0), dna.get(1), dna.get(2)
            d = dna.popfront()
            if d == 'C':
                p.append('I')
            elif d == 'F':
                p.append('C')
            elif d == 'P':
                p.append('F')
            elif d == 'I':
                d = dna.popfront()
                if d == 'C':
                    p.append('P')
                elif d == 'P':
                    n = nat(dna)
                    p.append('!' + str(n))
                elif d == 'F':
                    dna.popfront() # NOTE: Three bases consumed here.
                    s = consts(dna)
                    p.append('?' + ''.join(s))
                elif d == 'I':
                    d = dna.popfront()
                    if d == 'P':
                        lvl += 1
                        p.append('(')
                    elif d == 'C' or d == 'F':
                        if lvl > 0:
                            lvl -= 1
                            p.append(')')
                        else:
                            return p
                    elif d == 'I':
                        # Add rna command.                                                
                        rnacmd = dna.popfront(7)
##                        print rnacmd
                        if ''.join(rnacmd) not in rnacommands:
                            print 'PTN: Warning: Unknown RNA cmd: ' + ''.join(rnacmd)
                        rna.extend(rnacmd)
    except IndexError:
        pass
    raise NoMoreData

# Extracts a template from the list 'dna', starting at position 'pos'.
# Returns a tuple containing the template and a new position after the
# consumed bases: (template, pos)
# The 'rna' parameter is also modified during the process.
def template(dna, rna):
    t = []
    while dna:
        d = dna.popfront()
        if d == 'C':
            t.append('I')
        elif d == 'F':
            t.append('C')
        elif d == 'P':
            t.append('F')
        if d == 'I':
            d = dna.popfront()
            if d == 'C':
                t.append('P')
            elif d == 'F' or d == 'P':
                l = nat(dna)
                n = nat(dna)
                t.append((l, n))
            elif d == 'I':
                d = dna.popfront()
                if d == 'P':
                    n = nat(dna)
                    t.append(n)
                elif d == 'C' or d == 'F':
                    return t
                elif d == 'I':
                    # Add rna command.
                    rnacmd = dna.popfront(7)
##                    print rnacmd
                    if ''.join(rnacmd) not in rnacommands:
                        print 'Warning: Unknown RNA cmd: ' + ''.join(rnacmd)
                    rna.extend(rnacmd)
    raise NoMoreData

# Extracts a natural number from the list 'dna', starting at position 'pos'.
# Returns a tuple containing the number and a new position after the
# consumed bases: (number, pos)
def nat(dna):
    try:
        d = dna.popfront()
        if (d == 'P'):
            return 0
        elif (d == 'I') or (d == 'F'):
            n = nat(dna)
            return 2 * n
        elif (d == 'C'):
            n = nat(dna)
            return 2 * n + 1
        else:
            # Empty -> Exit
            raise NoMoreData
    except IndexError:
        raise NoMoreData
    
    
        
# Extracts a base sequence from the list 'dna', starting at position 'pos'.
# Returns a tuple containing the sequence and a new position after the
# consumed bases: (sequence, pos)
def consts(dna):
    def constsrec(dna):
        if not dna:
            return []
        d = dna.get(0)
        if d == 'C':
            dna.popfront()
            seq = consts(dna)
            seq.append('I')
            return seq
        elif d == 'F':
            dna.popfront()
            seq = constsrec(dna)
            seq.append('C')
            return seq
        elif d == 'P':
            dna.popfront()
            seq = constsrec(dna)
            seq.append('F')
            return seq
        elif d == 'I' and dna.get(1) == 'C':
            dna.popfront(2)
            seq = constsrec(dna)
            seq.append('P')
            return seq
        else:
            return []            
    seq = constsrec(dna)
    seq.reverse()
    return seq
    
# Modifies 'dna' by applying template 't' to matching items in pattern 'pat'.
# The matching starts at position 'i'.
def matchreplace(dna, pat, t):
    e = []
    c = []
    i = 0
##    print "matchreplace", dna, pat, t
    for p in pat:
        if p.startswith('!'):
            n = int(p[1:])
            i += n
            if (i > len(dna)):
                # Match failed.
##                print 'Matched failed in !'
                return
        elif p.startswith('?'):
            substr = p[1:]
            n = dna.find(substr, i)
            if n >= 0:
                i = n + len(substr)
            else:
                # Match failed.
##                print 'Matched failed in ?'
                return
        elif (p == '('):
            c.append(i)
        elif (p == ')'):
            s = (c.pop(), i)
            r = dna.getref(s)
            e.append(r)
        else:
            # Base
            if (dna.get(i) == p):
                i += 1
            else:
                # Match failed.
##                print 'Matched failed in Base', p
                return
    dna.popfront(i)
    replace(dna, t, e)
    
def replace(dna, tpl, e):
##    print 'Replace ', dna, tpl, e 
    r = []
    for t in tpl:
        if isinstance(t, int):
            # |n|
            if (t >= len(e)):
                r.extend(asnat(0))
            else:
                r.extend(asnat(len(e[t])))
        elif isinstance(t, tuple):
            # n(l)
            l, n = t
            if (n >= len(e)):
                r.extend(protect(l, []))
            else:
                r.extend(protect(l, e[n]))
        else:
            # Base
            r.append(t)
##    print "r:", r
    dna.prepend(r)
##    print "dna:", dna
    
def protect(l, d):
    while l:
        d = quote(d)
        l -= 1
    return d
        
def quote(d):
    nd = []    
    for item in d:
        if (item == 'I'):
            nd.append('C')
        elif (item == 'C'):
            nd.append('F')
        elif (item == 'F'):
            nd.append('P')
        else:
            # P
            nd.append('I')
            nd.append('C')
    return nd
    
def asnat(n):
    d = []
    while n > 0:
        if (n % 2) == 0:
            d.append('I')
        else:
            d.append('C')
        n /= 2
    d.append('P')
    return d
    
def execute(dna, rna, progress = False):
    pos = 0
    n = 0
    while True:
        n += 1
        try:
            if n == 75: print "hej"
            p = pattern(dna, rna)
            if n == 75: print "hej"
            t = template(dna, rna)
            if n == 75: 
                print "hej"
                print len(p), p, t
            matchreplace(dna, p, t)
            if progress:
                print 'Iterations: ' + str(n) + '   DNA remaining: ' + str(len(dna)), '   RNA commands: ' + str(len(rna) / 7)
        except NoMoreData:
            print 'DNA remaining: ' + str(len(dna))
            break

if __name__ == '__main__':
    prefix = ""
    if len(sys.argv) > 3:
        prefixfile = file(sys.argv[3], 'r')
        prefix = prefixfile.read()
        prefixfile.close()
    if len(sys.argv) > 2:
        dnafile = file(sys.argv[1], 'r')
        dna = DNAList(prefix + dnafile.read())
        dnafile.close()
        rna = []
        try:
            dna = execute(dna, rna, True)
        except KeyboardInterrupt:
            rnafile = file(sys.argv[2], 'w')
            rnafile.write(''.join(rna))
            rnafile.close()
            sys.exit(1)
        rnafile = file(sys.argv[2], 'w')
        rnafile.write(''.join(rna))
        rnafile.close()
        
    else:
        # Run tests
        print ''
        print 'Test pattern function:'
        print ''
        for dnastr in ['CIIC', 'IIPIPICPIICICIIF']:
            rna = []
            dna = DNAList(dnastr)
            p = pattern(dna, rna)
            print dnastr + ' -> ' + ''.join(p)
        print ''
        print ''
        print 'Test dna execution function:'
        print ''
        for dnastr in ['IIPIPICPIICICIIFICCIFPPIICCFPC', 'IIPIPICPIICICIIFICCIFCCCPPIICCFPC', 'IIPIPIICPIICIICCIICFCFC']:
            dna = DNAList(dnastr)
            rna = []
            execute(dna, rna)
            print dnastr + ' -> ' + ''.join(dna)

File: src/dna2rna_boost/dna2rna.cpp
#include "dnalist.hpp"
#include <vector>
#include <string>
#include <sstream>
#include <iostream>

bool dna_debug = false;
unsigned int dna_skip = 1;

class tmpl
{
public:
	tmpl(std::string s="", int l=-1, int n=-1) : s(s), l(l), n(n) {}
	std::string s;
	int l;
	int n;
};

typedef std::vector<tmpl> tvec;

int nat(DNAList& dna)
{
	char d = dna[0];
    dna.popfront(1);
    if (d == 'P')
	{
        return 0;
	}
    else if (d == 'I' || d == 'F')
	{
        return 2 * nat(dna);
	}
    else if (d == 'C')
	{
        return 1 + 2 * nat(dna);
	}
}    
    
std::string constsrec(DNAList& dna)
{
	char d = dna[0];		
	if (d == 'C')
	{
		dna.popfront(1);
		std::string seq = constsrec(dna);
		seq.push_back('I');
		return seq;
	}
	else if (d == 'F')
	{
		dna.popfront(1);
		std::string seq = constsrec(dna);
		seq.push_back('C');
		return seq;
	}
	else if (d == 'P')
	{
		dna.popfront(1);
		std::string seq = constsrec(dna);
		seq.push_back('F');
		return seq;
	}
	else if (d == 'I' && dna[1] == 'C')
	{
		dna.popfront(2);
		std::string seq = constsrec(dna);
		seq.push_back('P');
		return seq;
	}
	else
	{
		return "";
	}
}

std::string consts(DNAList& dna)
{
	std::string s = constsrec(dna);
	std::reverse(s.begin(), s.end());
	return s;
}


void patternfcn(DNAList& dna, svec& rna, svec& p)
{
	unsigned int lvl = 0;
    for (;;)
	{
		char d = dna[0];
        if (d == 'C')
		{
            dna.popfront(1);
            p.push_back("I");
		}
        else if (d == 'F')
		{
            dna.popfront(1);
            p.push_back("C");
		}
        else if (d == 'P')
		{
            dna.popfront(1);
            p.push_back("F");
		}
        else if (d == 'I')
        {
        	d = dna[1];
        	if (d == 'C')
			{
	            dna.popfront(2);
	            p.push_back("P");
			}
	        else if (d == 'P')
			{
	            dna.popfront(2);
	            int n = nat(dna);
				std::stringstream str;
				str << "!" << n;
				std::string s = str.str();
	            p.push_back(s);
			}
	        else if (d == 'F')
			{
	            dna.popfront(3); // NOTE: Three bases consumed here.
				std::string s("?");
				s.append(consts(dna));
				p.push_back(s);
			}
	        else if (d == 'I')
	        {
	        	d = dna[2];
	        	if  (d == 'P')
				{
		            dna.popfront(3);
		            ++lvl;
		            p.push_back("(");
				}
				else if (d == 'C' || d == 'F')
				{
		            dna.popfront(3);
		            if (lvl > 0)
					{
		                --lvl;
		                p.push_back(")");
					}
		            else
					{
		                break;
					}
				}
		        else if (d == 'I')
				{
					std::string rnacmd;
					dna.get<std::string>(rnacmd, 3, 10);
		            dna.popfront(10);
					rna.push_back(rnacmd);
				}
	        }
        }
	}
}


void templatefcn(DNAList& dna, svec& rna, tvec& t)
{
    for (;;)
	{
		char d = dna[0];
        if (d == 'C')
		{
            dna.popfront(1);
            t.push_back(tmpl("I"));
		}
        else if (d == 'F')
		{
            dna.popfront(1);
            t.push_back(tmpl("C"));
		}
        else if (d == 'P')
		{
            dna.popfront(1);
            t.push_back(tmpl("F"));
		}
        else if (d == 'I')
        {
        	d = dna[1];
        	if (d == 'C')
			{
	            dna.popfront(2);
	            t.push_back(tmpl("P"));
			}
	        else if (d == 'P' || d == 'F')
			{
	            dna.popfront(2);
	            int l = nat(dna);
	            int n = nat(dna);
	            t.push_back(tmpl("", l, n));
			}
	        else if (d == 'I')
	        {
	        	d = dna[2];
	        	if (d == 'P')
				{
		            dna.popfront(3);
		            int n = nat(dna);
					t.push_back(tmpl("", -1, n));
				}
				else if (d == 'C' || d == 'F')
				{
		            dna.popfront(3);
					break;
				}
		        else if (d == 'I')
				{
					std::string rnacmd;
					dna.get<std::string>(rnacmd, 3, 10);
		            dna.popfront(10);
					rna.push_back(rnacmd);
				}
	        }
        }
	}
}

void asnat(dnaseq& d, int n)
{
    while (n > 0)
	{
        if ((n % 2) == 0)
		{
            d.push_back('I');
		}
        else
		{
            d.push_back('C');
		}
        n /= 2;
	}
    d.push_back('P');
}


void quote(dnaseq& d)
{
	dnaseq nd;
	dnaseq::iterator it = d.begin();
    while (it != d.end())
	{
        if (*it == 'I')
		{
            nd.push_back('C');
		}
        else if (*it == 'C')
		{
			nd.push_back('F');
		}
        else if (*it == 'F')
		{
            nd.push_back('P');
		}
        else
		{
            // P
            nd.push_back('I');
            nd.push_back('C');
		}
		++it;
	}
	d.clear();
	std::copy(nd.begin(), nd.end(), std::back_inserter(d));
}


void protect(int l, dnaseq& d)
{
	if (l > 0)
	{
		while (l > 0)
		{
			quote(d);
			l -= 1;
		}
	}
}
        
void replacefcn(DNAList& dna, const tvec& tpl, evec& e, size_t i)
{
    dnareflist r;
	dnaseq* d = 0;
    for (tvec::const_iterator it = tpl.begin(); it != tpl.end(); ++it)
    {
    	const tmpl& t =*it; 
    	if (!t.s.empty())
    	{
            // Base
			if (d == 0)
			{
				d = dna.allocate();
			}
        	for (size_t ix = 0; ix < t.s.size(); ++ix)
        	{
        		d->push_back(t.s[ix]);
        	}
    	}
    	else if (t.l == -1)
        {
            // |n|
            if (t.n >= e.size())
            {
				if (d == 0)
				{
					d = dna.allocate();
				}
            	asnat(*d, 0);
			}
			else
			{
				if (d == 0)
				{
					d = dna.allocate();
				}
				std::pair<size_t, size_t>& a = e[t.n];
                asnat(*d, a.second-a.first);
            }
        }
        else
		{
            // n(l)
            if (t.n >= e.size())
            {
            	;
            }
            else
            {
                if (t.l == 0)
                {
					if (d)
					{
						r.push_back(new DNARef(0, d->size(), d));
						d = 0;
					}
                	std::pair<size_t, size_t>& a = e[t.n];
					dnareflist rl;
					dna.getreflist(rl, a.first, a.second);
					std::copy(rl.begin(), rl.end(), std::back_inserter(r));
                }
                else
                {
                	std::pair<size_t, size_t>& a = e[t.n];
					if (d)
					{
						r.push_back(new DNARef(0, d->size(), d));
					}
					d = dna.allocate();
					dna.get<dnaseq>(*d, a.first, a.second);
	            	protect(t.l, *d);
                }
            }
		}
    }
	if (d)
	{
		r.push_back(new DNARef(0, d->size(), d));
	}
    
	dna.popfront(i);
    dna.insertfront(r);
}

    


void matchreplace(DNAList& dna, const svec& pat, tvec& t)
{
    evec e;
    ivec c;
    size_t i = 0;
	if (dna_debug)
	{
		std::cout << "pattern ";
		for (svec::const_iterator it = pat.begin(); it != pat.end(); ++it)
		{
			std::string p = *it;
			std::cout << p;
		}
		std::cout << std::endl;
		std::cout << "template ";
		for (tvec::const_iterator it = t.begin(); it != t.end() && it != t.begin()+10; ++it)
		{
			tmpl tt = *it;
			if (tt.n == -1 && tt.l == -1)
			{
				std::cout << tt.s;
			}
			else
			{
				std::cout << "(" << tt.l << "," << tt.n << ")";
			}
		}	
		std::cout << "..." << std::endl;
	}
    for (svec::const_iterator it = pat.begin(); it != pat.end(); ++it)
	{
		std::string p = *it;
        if (p[0] == '!')
		{
            int n = atoi(p.substr(1).c_str());
            i += n;
            if (i > dna.size())
			{
				// Match failed.
                return;
			}
		}
        else if (p[0] == '?')
		{
			std::string substr = p.substr(1);
            size_t n = dna.find(substr, i);
            if (n >= 0)
			{
                i = n + substr.size();
			}
            else
			{
                // Match failed.
                return;
			}
		}
        else if (p[0] == '(')
		{
            c.push_back(i);
		}
        else if (p[0] == ')')
		{
			if (dna_debug)
			{
				std::cout << "e[" << e.size() << "] ";
				for (size_t ix = c.back(); ix < std::min(c.back()+10, i) ; ++ix)
				{
					std::cout << dna[ix];
				}
				std::cout << "... (" << i - c.back() << " bases)" << std::endl;
			}
            e.push_back(std::make_pair(c.back(), i));
			c.pop_back();
		}
        else
		{
            // Base
            if (dna[i] == p[0])
			{
                i += 1;
			}
            else
			{
                // Match failed.
                return;
			}
		}
	}
    replacefcn(dna, t, e, i);
}
    
  
void progress(DNAList& dna, svec& rna, unsigned int n)
{
	if (n % dna_skip == 0)
	{
		std::cout << "iteration " << n << std::endl;
		size_t sz = dna.size();
		std::cout << "dna = ";
		for (size_t i = 0; i < std::min(10u, sz); ++i)
		{
			std::cout << dna[i];
		}
		std::cout << "... (" << sz << " bases)" << std::endl;
		std::cout << "len(rna) == " << rna.size() << std::endl;
	}
}
    
void execute(DNAList& dna, svec& rna, bool prog, int iterations)
{
    unsigned int n = 0;
	if (prog)
	{
		progress(dna, rna, n);
	}
    for (;;)
	{
        ++n;
		svec p;
		tvec t;
		try
		{
			patternfcn(dna, rna, p);
			templatefcn(dna, rna, t);
			matchreplace(dna, p, t);
			if (prog)
			{
				progress(dna, rna, n);
			}
			if (iterations == n)
			{
				break;
			}
		}
		catch(int)
		{
			if (prog)
			{
				progress(dna, rna, n);
			}
			std::cout << "Finished!" << std::endl;
			break;
		}
	}
}

